### ISSUE TITLE
Degraded `elm repl` line editing experience on Debian and derivatives (Ubuntu, Linux Mint, ...)

### ISSUE THREAD

**Issue by rlefevre** [elm/compiler#1774]
### Summary
As `elm repl` uses [haskeline](http://hackage.haskell.org/package/haskeline), all its [key bindings](https://github.com/judah/haskeline/wiki/KeyBindings) should be available. However on Debian derivatives (based on [Debian versions >= 6](https://github.com/elm/compiler/issues/1768#issuecomment-417445263)), this is not the case. For example `CLTR-L` to clear the screen does not work, nor does `DELETE` to delete the character right of the cursor.

### Version
Elm 0.19.0 
Debian 9 / Ubuntu 18.04 / Linux Mint 19 (and all versions of these distributions since 2011)

### Severity
The issue does not prevent to use `elm repl` and does not seem to have any impact outside of it. It could most likely be classified as "minor" or "low".

### Impacted users
All users of a GNU/Linux distribution derived from Debian 6 (2011) or more are impacted, this includes users of Ubuntu and Linux Mint. According to [unreliable statistics](https://www.zdnet.com/article/whats-the-most-popular-linux-of-them-all/), this represents most likely the majority of Linux users.

### How to reproduce
* On a Debian derivative, run `elm repl`, then `CTRL-L`. This does not clear the screen as expected.
* On others Unix like systems, this can also be reproduced for tests purpose by setting an invalid `TERM` environment variable, for example by running `TERM=badterm elm repl`.

### Root cause
To work correctly in a given terminal, libraries like [`haskeline`](http://hackage.haskell.org/package/haskeline) or [`readline`](https://en.wikipedia.org/wiki/GNU_Readline) search on the system (through `libncurses/libtinfo`) the terminal description file with the same name as the `$TERM` environment variable (usually `xterm` or some variants) to detect the terminal capabilities. This file is expected to be located inside one of the directories listed in a [search path](https://www.decf.berkeley.edu/help/unix/searchpath.html) configured during the `ncurses` library build (using `--with-terminfo-dirs`). This directory, which is a database containing all terminal description files, is called `terminfo`.

The `elm` Linux binary is built on the Alpine Linux distribution that configures the terminfo search path during the `ncurses` library compilation as `/etc/terminfo:/usr/share/terminfo` whereas Debian distributions put the terminfo database in `/lib/terminfo` (and use `/etc/terminfo:/lib/terminfo:/usr/share/terminfo` for the search path). Consequently the terminal description file is not found on these Linux distributions.

### User configuration fixes
As the `ncurses` library uses several mechanisms to find the terminfo database, there are several configuration work-arounds to help it find it (see the annex or `man terminfo`). 

Any one of the following work-arounds should work:

1. Add a symbolic link to the terminfo database from the home directory:
```
ln -s /lib/terminfo ~/.terminfo
```
2. Start `elm` with the `TERMINFO` environment variable pointing to the correct directory on your system:
```
TERMINFO=/lib/terminfo elm repl
```
3. Export the `TERMINFO` variable in your `~/.bashrc` file:
```
export TERMINFO=/lib/terminfo
```
4. Export the `TERMINFO_DIRS` extension supported by `elm` in your `~/.bashrc` file:
```
export TERMINFO_DIRS=/etc/terminfo:/lib/terminfo:/usr/share/terminfo:/usr/lib/terminfo
```

### Possible permanent fixes

Based on the [analysis of Linux distributions](https://github.com/elm/compiler/issues/1768#issuecomment-417445263), [documentation](https://www.tldp.org/HOWTO/Text-Terminal-HOWTO-16.html) and [TERMINFO man page](https://linux.die.net/man/5/terminfo), the following set of directories to search for the terminfo database should work on most in not all systems:
```
/etc/terminfo:/lib/terminfo:/usr/share/terminfo:/usr/lib/terminfo
```
There are several options to help the `elm` binary using it:

1. Tell users to use one of the configuration fix listed in the previous section when appropriate. The issue is that it is not easy to detect the systems on which there is an issue without duplicating a part of the algorithm used by `ncurses` to find the terminfo database. Users may also be surprised and annoyed by this step as all their other readline/haskeline tools built by the distribution work without any additional configuration.

2. Rebuild the `ncurses` library with some additional terminfo directories and statically link the `elm` binary to it. ~The Alpine Linux [may include one day the needed directories](https://bugs.alpinelinux.org/issues/9362) but until then, the library would have to be rebuilt as part of the `elm` Linux binary build process.~ [Edit: the updated Alpine package has been released following our feature request, see [the following comment](https://github.com/elm/compiler/issues/1774#issuecomment-418171136)].  This solution requires to update the build process and presents a risk of regression if this process is later changed but has not impact whatsoever on other systems (MacOSX/Windows).

3. Configure one of the terminfo environment variable inside the `elm` binary before haskeline use. This is the solution explored by [this pull-request](https://github.com/elm/compiler/pull/1772) using the `TERMINFO_DIRS` environment variable, which is more or less an automatization of the user configuration fix 4. As this variable is used in complement of the other mechanisms to find the terminfo database and does not override any other configured directory, the risk of regression is low (see the annex below) but the code setting the variable would still run on systems where this is not needed. 

### Annex - algorithm used by ncurses to find the terminal description file
Excerpt from the [`TERMINFO(5)` man page](https://linux.die.net/man/5/terminfo):
```
The  ncurses library searches for terminal descriptions in several places.  It uses only the first description
found.  The library has a compiled-in list of places to search which can be overridden  by  environment  vari‐
ables.  Before starting to search, ncurses eliminates duplicates in its search list.

·   If  the  environment variable TERMINFO is set, it is interpreted as the pathname of a directory containing
    the compiled description you are working on.  Only that directory is searched.

·   If TERMINFO is not set, ncurses will instead look in the directory $HOME/.terminfo for a compiled descrip‐
    tion.

·   Next,  if the environment variable TERMINFO_DIRS is set, ncurses will interpret the contents of that vari‐
    able as a list of colon-separated directories (or database files) to be searched.

    An empty directory name (i.e., if the variable begins or ends with a colon, or contains  adjacent  colons)
    is interpreted as the system location /etc/terminfo.

·   Finally, ncurses searches these compiled-in locations:

    ·   a list of directories (no default value), and 

    ·   the system terminfo directory, /etc/terminfo (the compiled-in default).
```

**Comment by rlefevre** [comment]
The Alpine Linux package has been updated in `edge` (rolling release) with the following search path that fixes `elm` issue:
```
"/etc/terminfo:/usr/share/terminfo:/lib/terminfo:/usr/lib/terminfo"
```
See the following [commit](https://github.com/alpinelinux/aports/commit/e773ffa89a7b52f56f5cdbce5d3ab6518a192e8d)

So there are now two ways to fix the issue when building the next elm release (using possible fix 2. in previous comment):

### 1. Build on Alpine 3.7 by just importing the `ncurses` packages from `edge` 
This can be done with this modification in the `Dockerfile`:
```
+# Install edge version of ncurses with fixed terminfo search dir
+RUN apk add ncurses-dev ncurses-static --update-cache --repository http://dl-3.alpinelinux.org/alpine/edge/main/
+
 # Install required packages
-RUN apk add --update ghc cabal git musl-dev zlib-dev ncurses-dev ncurses-static
+RUN apk add --update ghc cabal git musl-dev zlib-dev
```

This modification can be removed later once a new Alpine version has been released and used to build `elm`.

### 2. Build `elm` with the `edge` version of Alpine (or next stable version once released):

* This requires first to fix the build issue with `ghc-8.4` as Alpine `edge` includes `ghc-8.4.3`: https://github.com/elm/compiler/pull/1784
* Then just changing the release in the `Dockerfile` and adding `wget` in the packages list is enough (`wget` is now required by `cabal`):
```
-FROM alpine:3.7
+FROM alpine:edge
```
```
 # Install required packages
-RUN apk add --update ghc cabal git musl-dev zlib-dev ncurses-dev ncurses-static
+RUN apk add --update ghc cabal git musl-dev zlib-dev ncurses-dev ncurses-static wget

```

As the fixed `ncurses` package will be included in the next Alpine stable release, this release can be used instead of `edge` once released (most likely `3.8.1`)

**Comment by rlefevre** [comment]
Edit: The fixed Alpine package has been [backported to Alpine 3.8](https://git.alpinelinux.org/cgit/aports/commit/?id=e3d3bc4817ac56d96ce5d2c4c891e431070a766e) stable release. I updated the previous comment consequently.

**Comment by stevensonmt** [comment]
This issue is not limited to Debian based distros as I have it on my Manjaro system as well.

**Comment by HappMacDonald** [comment]
I see it on Cygwin too.
To be fair, I get same error on irb (ruby) and iex (elixir) read/evaluate/print/loop tools as well so it's not clear to me if fault is best assigned to elm-compiler or to the environment? :J

**Comment by rlefevre** [comment]
I think that this bug is fully fixed on Linux in elm 0.19.1 (by building with Alpine 3.10 which includes our PR).

**Comment by HappMacDonald** [comment]
Cool, I can confirm repl readline behaves itself in Ubuntu linux, now.

Cygwin behavior appears unchanged. f/e pressing "up" arrow just moves the mouse up the screen, and gets interpreted by repl as a bunch of nonsense characters in the input stream.

On a mild tangent, I also think it's worth noting that the Windows Installer trips a relatively worrisome trigger on [virustotal](https://www.virustotal.com/gui/file/02947d00090f556abcf8f8e95997ac993825bd783abf05786afe8f60a002d496/detection):
> McAfee-GW-Edition: BehavesLike.Win32.Ransom.wc

Thanks guys. :D

**Comment by rlefevre** [comment]
@HappMacDonald It may be better to open another issue specific to Cygwin as this bug was about Linux and could be closed. Same for the virustotal thing.

**Comment by rlefevre** [comment]
Elm 0.19.1 Linux binary has been [built with Alpine 3.10](https://github.com/elm/compiler/blob/c9aefb6230f5e0bda03205ab0499f6e4af924495/installers/linux/Dockerfile#L4).

This Alpine release [includes](https://gitlab.alpinelinux.org/alpine/aports/commit/e3d3bc4817ac56d96ce5d2c4c891e431070a766e) our [feature request](https://gitlab.alpinelinux.org/alpine/aports/issues/9362) to add `/lib/terminfo` and `/usr/lib/terminfo` to the searched directory for the terminal description that was not found in 0.19.0 on Debian derivatives and other distributions that store it there.

Consequently this bug should be fixed for all Linux distributions listed in this issue for Elm releases >= 0.19.1.

If you still have some issues with some `elm repl` key bindings on another OS or another Linux distribution, please open a new issue.
